rfcs/0000-plan-dynanism.md

```diff
@@ -0,0 +1,251 @@
+To this new generation of developers, the distro (or homebrew) is a crufty relic from an earlier age to bootstrap modernity, and then be forgotten about.
+
+Right now, to deal with these packages, we either convert by hand, or commit lots of generated code into Nixpkgs.
+But I don't think either of those options are healthy or sustainable.
+The problem with the first is sheer effort; we'll never be able to keep up.
+The problem with the second is bloating Nixpkgs but more importantly reproducability: If someone wants to update that generated code it is unclear how.
+All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.
+
+The core feature here, derivations that build derivations, is the best fundamental primitive for this problem.
+It's very performant, being well-adapted for Nix's current scheduler.
+Unlike recursive Nix, there's is no potential for half-built dependencies to sit around waiting for other builds, wasting resources.
+Each build step (derivation) always runs start to finish blocking on nothing.
+It's very efficient, because it doesn't obligate the use of the Nix expression language.
+Finally, it's quite compatible with `--dry-run`.
+
+However, "import from derivation" is still far and away the easiest method to use, and the one that existing tools to convert to Nix use.
+We should continue to support it, finding a way for `hydra.nixos.org` to allow it, so those tools with IFD can be used in Nixpkgs and become first-class members of the Nix ecosystem.
+We have a fairly straightforward mechanism, only slightly more cumbersome than IFD today, to allow evaluation to be deferred after imported things are built.
+This frees up the Hydra evaluator to finish before building, and also meshes well with any plan to build more than one eval-realized path at a time.
+This should allow us to drop the `hydra.nixos.org` restriction.
+
+With these steps, I think we will be able to successfully convert to Nix a bunch of developers that mainly work in one language, and didn't even think they were in need of a better distro.
+In turn, I hope these upstream packages and ecosystems might even care about packaging and integration of the sort that we do.
+This would create a virtuous cycle where Nix is easier to use by more people, and Nixpkgs is easier to maintain as upstream packages better match our values.
+
+# Detailed design
+[design]: #detailed-design
+
+We can break this down nicely into steps.
+
+## Dynamic derivations
+
+*This is implemented in https://github.com/NixOS/nix/pull/4628.*
+
+1. Derivation outputs can be valid derivations:
+
+   1. Allow derivation outputs to be content addressed with the "text hashing" scheme.
+
+   2. Lift the restriction barring derivations output paths from ending in `.drv` if they are so content-addressed
+
+2. Extend the CLI to take advantage of such derivations:
```

Comment:
Specifying the CLI here feels a bit odd. Shouldn't this feature, at first, be transparent to the CLI? If someone uses the new hydra-safe IFD that shouldn't matter to the CLI.

In general overloading the "installables" attribute more and more (first flakes with `#` and `!` and introducing `*`) feels very weird. They might be safe on posix shells but they are to me not the obvious way of handling these situations.

Can't we be explicit about the intentions?

```
# build the out output
$ nix build -f /nix/store/something-....drv out

# build the man output
$ nix build -f /nix/store/something-....drv man

# build the foo attribute of the "default" flake/expression in the current directory
$ nix build foo

# build bar.man from the default.nix in the current dir
$ nix build -f default.nix bar.man


# build the pythonPackages.black attribute of nixpkgs
$ nix build -f channel:nixos-unstable pythonPacakges.black
```

This would at least keep a clear separation of attributes and expression/derivation to build without adding more and more complexity into "installables".

Reply:
So Flake's `#` is purely a higher level concern. My thing however allows a fully faithful representation of today's' `DerivedPath`, which is a solidly libnixstore concept corresponding to the sorts of top level "goals" we have in the schedule.

Here's another way to think about it: what would
```bash
# build the out output
$ nix path-info /nix/store/something-....drv out
```
mean with your proposal?
`nix build -f /nix/store/something-....drv out` looks (to me) like this might be a `nix build` only thing, but mine is supposed to make clear that `/nix/store/something-....drv!out` is one concept regardless of whether we are building, getting the path info, or any other operation: it's purposefully a new type of argument to all command not new type of command.

Reply:
Even more concretely `nix path-info /nix/store/something-....drv!out` would tell you about a realization, which may or may not be filled in yet by substitution or building, and even if it does exist may or may not point to a path one actually has.

I hope that this example shows how `/nix/store/something-....drv!out` is a useful concept for many different actions, which is why I want to make it a "first class" argument type.

Reply:
On the topic of `!`: I would find both `@`  and `~` more intuitive, since they have come to represent a pointer to something (a user or a host), which seems to match what we're doing here/